File: src/AllOfRule.java
import java.util.List;
public class AllOfRule implements RequirementRule {
    private List<RequirementRule> rules;

    public AllOfRule(List<RequirementRule> rules) {
        this.rules = rules;
    }

    @Override
    public boolean isSatisfied(Student student) {
        for (RequirementRule rule : rules) {
            if (!rule.isSatisfied(student)) {
                return false;
            }
        }
        return true;
    }
}

File: src/all_of_rule.cpp
#include "all_of_rule.hpp"

#include "student.hpp"

#include <utility>

namespace course_planner {

AllOfRule::AllOfRule(std::vector<RequirementRulePtr> rules)
    : rules_(std::move(rules)) {}

bool AllOfRule::is_satisfied(const Student& student) const {
    for (const auto& rule : rules_) {
        if (!rule->is_satisfied(student)) {
            return false;
        }
    }
    return true;
}

}  // namespace course_planner
